pennylane_lightning/core/src/simulators/lightning_kokkos/catalyst/LightningKokkosSimulator.cpp

```diff
@@ -342,13 +342,22 @@ void LightningKokkosSimulator::PartialProbs(
     std::move(dv_probs.begin(), dv_probs.end(), probs.begin());
 }
 
-void LightningKokkosSimulator::Sample(DataView<double, 2> &samples,
-                                      std::size_t shots) {
+std::vector<size_t> LightningKokkosSimulator::GenerateSamples(const size_t shots) {
+    // generate_samples is a member function of the Measures class.
     Pennylane::LightningKokkos::Measures::Measurements<StateVectorT> m{
         *(this->device_sv)};
+
     // PL-Lightning-Kokkos generates samples using the alias method.
     // Reference: https://en.wikipedia.org/wiki/Inverse_transform_sampling
-    auto li_samples = m.generate_samples(shots);
+    if (this->gen) {
+        return m.generate_samples(shots, (*(this->gen))());
+    }
+    return m.generate_samples(shots);
```

Comment:
Question to the C++ experts, would an additional function like this (as opposed to calling `m.generate_samples` directly in the caller) ever lead to additional copies (from the local object in this scope to the object in the caller scope), or do optimizations pretty much guarantee this doesn't happen?

Reply:
(Note that [lightning.qubit in catalyst has this helper function](https://github.com/PennyLaneAI/catalyst/blob/934726fe750043886415953dbd89a4c4ddeb9a80/runtime/lib/backend/lightning/lightning_dynamic/LightningSimulator.cpp#L302); not sure if the performance folks inlined it back when porting kokkos)

Reply:
Ah, that function even explicitly mentions return value optimization!

Reply:
Usually copy elision should come for free here, so should be all good (https://en.cppreference.com/w/cpp/language/copy_elision the section on prvalue semantics here should match the above case).